### PR TITLE
Add put index template api to high level rest client

### PR DESCRIPTION
**Pull request by dnhatn**
Relates #27205

### CODE REVIEW

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java [hunk 1]
```diff
@@ -662,6 +662,15 @@ public static String randomRealisticUnicodeOfCodepointLength(int codePoints) {
         return generateRandomStringArray(maxArraySize, maxStringSize, allowNull, true);
     }
 
+    public static List<String> randomStringList(int minListSize, int maxListSize, int minStringLength, int maxStringLength) {
```
Comment (dakrone) [inline test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java]: Can you add javadocs for this please?

Comment (dakrone) [reply test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java]: Also, do we use the `minStringLength` and `maxStringLength`? Just wondering if instead we could do `Arrays.asList(generateRandomStringArray(...))`, unless we need the unicode-ness and the min/max string size

**Review by dakrone — APPROVED** [review]
LGTM, I left one question/comment but it's nothing required

File: test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java (hunk 1 above)
Comment (dnhatn) [reply test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java]: Yeah, I removed the new method and used `generateRandomStringArray`.

**Comment by dnhatn** [comment]
Thanks @dakrone for reviewing!

File: server/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequest.java [hunk 2]
```diff
@@ -539,4 +540,34 @@ public void writeTo(StreamOutput out) throws IOException {
...
+        }
+
+        builder.startObject("settings");
+        settings.toXContent(builder, params);
+        builder.endObject();
+
+        builder.startObject("mappings");
+        for (Map.Entry<String, String> entry : mappings.entrySet()) {
+            Map<String, Object> mapping = XContentHelper.convertToMap(new BytesArray(entry.getValue()), false).v2();
+            builder.field(entry.getKey(), mapping);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequest.java]: could of things here: we should not use a deprecated method, I think it is guaranteed that the request holds json (we convert it on the setter methods), hence we should not need content-type auto-detection. Furthermore, we don't need to parse stuff back into a map to the write it back to the builder. I would try and do something like this:

```
builder.field(entry.getKey());
XContentParser parser = JsonXContent.jsonXContent.createParser(NamedXContentRegistry.EMPTY, DeprecationHandler
                    .THROW_UNSUPPORTED_OPERATION, entry.getValue());
builder.copyCurrentStructure(parser);
```

File: server/src/test/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequestTests.java [hunk 3]
```diff
@@ -131,4 +137,52 @@ public void testValidateErrorMessage() throws Exception {
         assertThat(noError, is(nullValue()));
     }
 
+    private PutIndexTemplateRequest randomPutIndexTemplateRequest() throws IOException {
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequestTests.java]: can this be static?

File: server/src/test/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequestTests.java [hunk 4]
```diff
@@ -131,4 +137,52 @@ public void testValidateErrorMessage() throws Exception {
...
+            BytesReference shuffled = toShuffledXContent(expected, xContentType, ToXContent.EMPTY_PARAMS, randomBoolean());
+            PutIndexTemplateRequest parsed = new PutIndexTemplateRequest().source(shuffled, xContentType);
+            assertNotSame(expected, parsed);
+            assertThat(parsed.version(), equalTo(expected.version()));
+            assertThat(parsed.order(), equalTo(expected.order()));
+            assertThat(parsed.patterns(), equalTo(expected.patterns()));
+            assertThat(parsed.aliases(), equalTo(expected.aliases()));
+            assertThat(parsed.mappings(), equalTo(expected.mappings()));
+            assertThat(parsed.settings(), equalTo(expected.settings()));
+        }
```
Comment (javanna) [inline server/src/test/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateRequestTests.java]: I would make this class extend `AbstractXContentTestCase`, then your randomPutIndexTemplateRequest would become `createTestInstance`, and

```
    @Override
    protected PutIndexTemplateRequest doParseInstance(XContentParser parser) throws IOException {
        return new PutIndexTemplateRequest().source(parser.map());
    }

    @Override
    protected boolean supportsUnknownFields() {
        return false;
    }

    @Override
    protected void assertEqualInstances(PutIndexTemplateRequest expected, PutIndexTemplateRequest parsed) {
        assertNotSame(expected, parsed);
        assertThat(parsed.version(), equalTo(expected.version()));
        assertThat(parsed.order(), equalTo(expected.order()));
        assertThat(parsed.patterns(), equalTo(expected.patterns()));
        assertThat(parsed.aliases(), equalTo(expected.aliases()));
        assertThat(parsed.mappings(), equalTo(expected.mappings()));
        assertThat(parsed.settings(), equalTo(expected.settings()));
    }
```

File: server/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateResponse.java [hunk 5]
```diff
@@ -47,4 +49,14 @@ public void writeTo(StreamOutput out) throws IOException {
...
     }
+
+    private static final ConstructingObjectParser<PutIndexTemplateResponse, Void> PARSER;
+    static {
+        PARSER = new ConstructingObjectParser<>("put_index_template", true, args -> new PutIndexTemplateResponse((boolean) args[0]));
+        declareAcknowledgedField(PARSER);
+    }
+
+    public static PutIndexTemplateResponse fromXContent(XContentParser parser) {
+        return PARSER.apply(parser, null);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/admin/indices/template/put/PutIndexTemplateResponse.java]: Although this response is super simple, we may want to add a PutIndexTemplateResponseTests similar to UpdateSettingsResponseTests ?

**Comment by javanna** [comment]
heya @dnhatn thanks a lot for working on this, I left a few minor comments.

**Comment by dnhatn** [comment]
Hi @javanna, I will address your comments in a follow-up.

**Pull request merged**